internals/daemon/daemon.go

```diff
@@ -193,6 +200,19 @@ func (c *Command) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
+	// Optimisation: avoid calling userFromRequest, which acquires the state
+	// lock, in case we don't need to (when endpoint is OpenAccess). This
+	// avoids holding the state lock for /v1/health in particular, which is
+	// not good: https://github.com/canonical/pebble/pull/369
+	var user *UserState
+	if _, isOpen := access.(OpenAccess); !isOpen {
+		user, err = userFromRequest(c.d.state, r, ucred)
+		if err != nil {
+			Forbidden("forbidden").ServeHTTP(w, r)
+			return
+		}
+	}
+
```

Comment:
Open questions to discuss with @hpidcock:

After the `access.CheckAccess()` call and before `rspf()`, do we want to do something like this, to make it easier for (the few) handlers that behave differently based on access level, like notices?

This way the `UserState` value passed to the handler would be present for named identity-based users, as well as "default" users (like uid 0/root and uid==daemon_uid).

```
  if user == nil && ucred != nil {
    if ucred.Uid == 0 || ucred.Uid == uint32(os.Getuid()) {
      user = &UserState{Access: state.AdminAccess}
    } else {
      user = &UserState{Access: state.ReadAccess}
    }
  }

  rsp := rspf(c, r, user)
```

Could also add `Ucred *ucred` or `Uid *uint32` to `UserState` to make it easier for notices handlers to get the uid, instead of re-parsing it from the request with `ucrednetGet`.

Reply:
TODO: That reminds me, if we don't do the above, I also need to update the notices code to handle named identity-based users (that is `UserState` not nil).

Reply:
Per discussion with Harry, I've made this change now, even when calling `CheckAccess()`, so that all the messy ucred stuff is done in daemon, and CheckAccess and handlers don't have to worry about it. I've also included UID in UserState so notices handlers (for example) don't have to worry about it.